File: sections/34_probability_expected_value/problems/c_coupon_collector/solution.cpp
#include <bits/stdc++.h>
using namespace std;

const long long MOD = 1000000007LL;

long long mod_pow(long long a, long long b) {
    long long result = 1;
    while (b > 0) {
        if (b & 1LL) result = result * a % MOD;
        a = a * a % MOD;
        b >>= 1LL;
    }
    return result;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int q;
    if (!(cin >> q)) return 0;
    vector<int> values(q);
    int limit = 0;
    for (int &n : values) {
        cin >> n;
        limit = max(limit, n);
    }
    vector<long long> harmonic(limit + 1, 0);
    for (int i = 1; i <= limit; i++) {
        harmonic[i] = (harmonic[i - 1] + mod_pow(i, MOD - 2)) % MOD;
    }
    for (int n : values) cout << n * harmonic[n] % MOD << '\n';
    return 0;
}

File: sections/34_probability_expected_value/problems/c_coupon_collector/solution.py
import sys


MOD = 1_000_000_007


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    q = data[0]
    values = data[1 : 1 + q]
    limit = max(values)
    inv = [0] * (limit + 1)
    harmonic = [0] * (limit + 1)
    for i in range(1, limit + 1):
        inv[i] = pow(i, MOD - 2, MOD)
        harmonic[i] = (harmonic[i - 1] + inv[i]) % MOD
    print("\n".join(str(n * harmonic[n] % MOD) for n in values))


if __name__ == "__main__":
    main()
